### PR TITLE
planner, executor: push limit down into IndexLookUpReader executor

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/10668

### What is changed and how it works?

Sink `PhysicalLimit` on top of `IndexLookUpReader` into `IndexLookUpReader`, and in executor, skip the preceding and tailing handles returned by `IndexPlan`. This optimization can avoid unnecessary `TableScan` for handles not satisfying `Limit / Offset` constraint, and improve execution performance especially for queries with large `Offset`.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported variable/fields change

Side effects

 - Increased code complexity

Related changes

 - Need to update cost of `IndexLookUpReader`'s `TablePlan / IndexPlan` since some unnecessary I/O and network cost is avoided. This can be done in the next PR.

### CODE REVIEW

**Comment by jingyugao** [comment]
This pr seems to do some query-rewrite in cbo pahse?

**Comment by eurekaka** [comment]
@jingyugao Yes, it is query-rewrite to some extent, like the operator pushdown (to storage layer) here.

File: executor/distsql.go [hunk 1]
```diff
@@ -679,20 +684,43 @@ func (w *indexWorker) fetchHandles(ctx context.Context, result distsql.SelectRes
...
+	handles []int64, retChk *chunk.Chunk, scannedKeys uint64, err error) {
 	handleOffset := chk.NumCols() - 1
 	handles = make([]int64, 0, w.batchSize)
+	// PushedLimit would always be nil for CheckIndex or CheckTable, we add this check just for issurance.
+	checkLimit := (w.PushedLimit != nil) && (w.checkIndexValue == nil)
 	for len(handles) < w.batchSize {
-		chk.SetRequiredRows(w.batchSize-len(handles), w.maxChunkSize)
+		requiredRows := w.batchSize - len(handles)
+		if checkLimit {
+			leftCnt := int(w.PushedLimit.Offset + w.PushedLimit.Count - scannedKeys - count)
```
Comment (lamxTyler) [inline executor/distsql.go]: Will it cause overflow when covert from uint64 to int?

File: executor/distsql.go [hunk 2]
```diff
@@ -484,9 +487,9 @@ func (e *IndexLookUpExecutor) startIndexWorker(ctx context.Context, kvRanges []k
 		}
 		if e.runtimeStats != nil {
 			copStats := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRootStats(e.idxPlans[len(e.idxPlans)-1].ExplainID().String())
-			copStats.SetRowNum(count)
+			copStats.SetRowNum(int64(count))
```
Comment (francis0407) [inline executor/distsql.go]: Not related to this PR, but it seems really strange to use `int64` instead of `uint64` for a `count`.

Comment (eurekaka) [reply executor/distsql.go]: @qw4990 Is there any specific reason why we are using `int`/`int64` for `SetRequiredRows`/`SetRowNum` respectively?

File: planner/core/task.go [hunk 3]
```diff
@@ -581,9 +582,42 @@ func (p *PhysicalLimit) attach2Task(tasks ...task) task {
...
+		if !isDoubleRead {
+			return false
+		}
+	}
+	// We can sink Limit into IndexLookUpReader only if tablePlan contains no Selection.
+	ts, isTableScan := reader.tablePlan.(*PhysicalTableScan)
+	if !isTableScan {
+		return false
+	}
+	reader.PushedLimit = &PushedDownLimit{
```
Comment (zz-jason) [inline planner/core/task.go]: can we push this `limit` as a child of the the index scan side? i.e. after pushdown, the plan may looks like:
```
nested loop join(root)
  limit(root)
    index reader(root)
      limit(cop)
  table reader(cop)
```

Comment (eurekaka) [reply planner/core/task.go]: We have already made this before this PR?

File: executor/distsql.go (hunk 2 above)
Comment (qw4990) [reply executor/distsql.go]: For `RowNum`, we use methods like `atomic.XXXInt64` to maintain it so its type is `int64`.
And for `RequiredRows` we maintain it using Golang statement and for convenience, we set its type to `int`.

**Pull request merged**